source/common/router/scoped_rds.cc

```diff
@@ -0,0 +1,272 @@
+#include "common/router/scoped_rds.h"
+
+#include <memory>
+
+#include "envoy/admin/v2alpha/config_dump.pb.h"
+#include "envoy/api/v2/srds.pb.validate.h"
+
+#include "common/common/assert.h"
+#include "common/config/subscription_factory.h"
+
+// Types are deeply nested under Envoy::Config::ConfigProvider; use 'using-directives' across all
+// ConfigProvider related types for consistency.
+using Envoy::Config::ConfigProvider;
+using Envoy::Config::ConfigProviderInstanceType;
+using Envoy::Config::ConfigProviderManager;
+using Envoy::Config::ConfigProviderPtr;
+
+namespace Envoy {
+namespace Router {
+
+ConfigProviderPtr ScopedRoutesConfigProviderUtil::maybeCreate(
+    const envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager&
+        config,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+    ConfigProviderManager& scoped_routes_config_provider_manager) {
+  if (config.route_specifier_case() != envoy::config::filter::network::http_connection_manager::v2::
```

Comment:
Should this be a contract with caller and `ASSERT`?

Reply:
Done.